### PR TITLE
feature/PATRO23-193 - create user in kc

### PR DESCRIPTION
**Pull request by IzabelaPalubicka**
https://tracker.intive.com/jira/browse/PATRO23-193

### CODE REVIEW

File: src/api/app/Controllers/SignUpController.cs [hunk 1]
```diff
@@ -1,7 +1,7 @@
 using FluentValidation;
 using Intive.Patronage2023.Api.User.CreatingUser;
 using Intive.Patronage2023.Modules.Example.Application.Example;
-using Intive.Patronage2023.Shared.Abstractions.Commands;
+using Intive.Patronage2023.Shared.Abstractions.Queries;
 using Microsoft.AspNetCore.Authorization;
 using Microsoft.AspNetCore.Mvc;
```
Comment (DawidBski) [inline src/api/app/Controllers/SignUpController.cs:7]: Wrong namespace

**Review by DawidBski — CHANGES_REQUESTED** [review]

**Review by Kepeczsz — APPROVED** [review]
For me it looks okay, after changing namespace like dawid suggested (I didn't change from my branch, so it's not even your fault)

**Review by DawidBski — APPROVED** [review]

File: src/api/app/Controllers/SignUpController.cs [hunk 2]
```diff
@@ -14,33 +14,33 @@ namespace Intive.Patronage2023.Modules.Example.Api.Controllers;
 [Route("[controller]")]
 public class SignUpController : ControllerBase
 {
-	private readonly ICommandBus commandBus;
+	private readonly IQueryBus queryBus;
```
Comment (pzajaczkowski) [inline src/api/app/Controllers/SignUpController.cs:17]: I am not sure about changing it to query. We do not return any value or search for it but create new account in our system. But my thinking may be wrong.

File: src/api/app/Controllers/SignUpController.cs [hunk 3]
```diff
@@ -14,33 +14,33 @@ namespace Intive.Patronage2023.Modules.Example.Api.Controllers;
...
 	private readonly IValidator<CreateUser> createUserValidator;
 
 	/// <summary>
 	/// Initializes a new instance of the <see cref="SignUpController"/> class.
 	/// </summary>
-	/// <param name="commandBus">command bus.</param>
+	/// <param name="queryBus">command bus.</param>
```
Comment (pzajaczkowski) [inline src/api/app/Controllers/SignUpController.cs:23]: If u decide o change to query, adjust comments

File: src/api/app/Controllers/SignUpController.cs [hunk 4]
```diff
@@ -14,33 +14,33 @@ namespace Intive.Patronage2023.Modules.Example.Api.Controllers;
...
-		this.commandBus = commandBus;
+		this.queryBus = queryBus;
 		this.createUserValidator = createUserValidator;
 	}
 
 	/// <summary>
 	/// Creates user.
 	/// </summary>
-	/// <param name="command">Command.</param>
+	/// <param name="query">Command.</param>
```
Comment (pzajaczkowski) [inline src/api/app/Controllers/SignUpController.cs:34]: Same as above

File: src/api/app/appsettings.Development.json [hunk 5]
```diff
@@ -15,6 +15,9 @@
 		"verify-token-audience": false,
 		"ssl-required": "none",
 		"resource": "inbudget-client",
+		"credentials": {
+			"secret": "0qX88ALVE2Xx9xeV9JKo5aCOdilNHx05"
+		},
 		"public-client": true,
```
Comment (pzajaczkowski) [inline src/api/app/appsettings.Development.json:21]: If we are using secret and authenticate the client, it don't has to be public anymore imo

**Review by pzajaczkowski — CHANGES_REQUESTED** [review]
One more thing. Could you clean up our signup/signin structure a bit?

image

I think there should be similar url (/SignUp /SignIn), and userController could be renamed to SignInController etc. SignIn is already on dev so it is the best place to adjust those things :)

File: src/api/app/Controllers/SignUpController.cs [hunk 6]
```diff
@@ -50,12 +50,12 @@ public SignUpController(ICommandBus commandBus, IValidator<CreateUser> createUse
...
 	[HttpPost]
-	public async Task<IActionResult> SignUp([FromBody] CreateUser command)
+	public async Task<IActionResult> SignUp([FromBody] CreateUser query)
 	{
-		var validationResult = await this.createUserValidator.ValidateAsync(command);
+		var validationResult = await this.createUserValidator.ValidateAsync(query);
 		if (validationResult.IsValid)
 		{
-			await this.commandBus.Send(command);
+			await this.queryBus.Query<CreateUser, HttpResponseMessage>(query);
```
Comment (mateuszmuszynski) [inline src/api/app/Controllers/SignUpController.cs:58]: We should use queries for scenarios where we would like to return data. We should use command in this case.

File: src/api/app/Keycloak/KeycloakService.cs [hunk 7]
```diff
@@ -43,16 +46,85 @@ public async Task<HttpResponseMessage> SignInGetToken(string email, string passw
...
+	{
+		string? resource = this.apiKeycloakSettings?.Resource;
+		string? realm = this.apiKeycloakSettings?.Realm;
+		string? secret = this.apiKeycloakSettings?.Credentials?.Secret;
+
+		string? url = $"/realms/{realm}/protocol/openid-connect/token";
+
+		var content = new FormUrlEncodedContent(new[]
+		{
+			new KeyValuePair<string, string>("client_id", resource ?? string.Empty),
```
Comment (mateuszmuszynski) [inline src/api/app/Keycloak/KeycloakService.cs:80]: It doesn't make sense to pass empty strings to http request content.

File: src/api/app/Keycloak/KeycloakService.cs [hunk 8]
```diff
@@ -43,16 +46,85 @@ public async Task<HttpResponseMessage> SignInGetToken(string email, string passw
...
 	}
+
+	/// <summary>
+	/// Get client token method.
+	/// </summary>
+	/// <param name="cancellationToken">A cancellation token that can be used to cancel the request.</param>
+	/// <returns>HttpResponseMessage object.</returns>
+	public async Task<HttpResponseMessage> GetClientToken(CancellationToken cancellationToken)
+	{
+		string? resource = this.apiKeycloakSettings?.Resource;
```
Comment (mateuszmuszynski) [inline src/api/app/Keycloak/KeycloakService.cs:72]: From a business point of view it doesn't make any sense to allow nulls here. This configuration MUST exist.

**Review by mateuszmuszynski — CHANGES_REQUESTED** [review]

**Review by pzajaczkowski — APPROVED** [review]
Haven't tested but it looks fine

**Review by mateuszmuszynski — APPROVED** [review]

**Pull request merged**